port/osx/omrosbacktrace_impl.c

```diff
@@ -0,0 +1,431 @@
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+/**
+ * @file
+ * @ingroup Port
+ * @brief Stack backtracing support
+ */
+#ifndef _GNU_SOURCE
+#define _GNU_SOURCE
+#endif
+
+#include "omrport.h"
+#include "omrportpriv.h"
+#include "omrsignal_context.h"
+
+#include <dlfcn.h>
+#include <execinfo.h>
+#include <stdlib.h>
+#include <string.h>
+#define UNW_LOCAL_ONLY
+#include <libunwind.h>
+
+#include "omrintrospect.h"
+
+static uintptr_t protectedBacktrace(OMRPortLibrary *port, void *arg);
+static uintptr_t backtrace_sigprotect(OMRPortLibrary *portLibrary, J9PlatformThread *threadInfo, void **address_array, uintptr_t capacity);
+
+typedef struct frameData {
+	void **address_array;
+	uintptr_t capacity;
+} frameData;
+
+/**
+ * Handles segfaults where either an invalid address is called or a signal was received in kernel mode.
+ * Both cases leave the ip register address invalid and libunwind fails to iterate past _sigtramp. This
+ * function fetches previous frame information and restores correct register values to the unwind cursor.
+ * It also records the stack frame addresses of the remaining frames.
+ *
+ * @param[in] cursor keeps track of stack unwind state
+ * @param[out] array contains the backtrace addresses to be returned
+ * @param[in] size size of the array
+ * @param[in] nAddr number of addresses currently recorded in the array
+ *
+ * @return total number of addresses recorded in the array
+ */
+static uintptr_t
+handle_sigtramp(unw_cursor_t *cursor, void **array, uintptr_t size, uintptr_t nAddr)
+{
+#if defined(OMR_ARCH_X86)
+	unw_word_t prevFrame = 0;
+	unw_word_t ip = 0;
+	ucontext_t *uc = NULL;
+	OMRUnixSignalInfo info;
+	const char *name = NULL;
+	void *reg = NULL;
+
+	/* ucontext for frame prior to _sigtramp is stored in rbx. Fetch previous frame info from it.
+	 * See https://opensource.apple.com/source/Libc/Libc-498/x86_64/sys/_sigtramp.s.auto.html
+	 */
+	unw_get_reg(cursor, UNW_X86_64_RBX, &prevFrame);
+	uc = (ucontext_t *)prevFrame;
+	info.platformSignalInfo.context = uc;
+
+	/* Copy the necessary machine regs to cursor so we can iterate the previous frame.
+	 * See https://opensource.apple.com/source/libunwind/libunwind-35.4/src/unw_getcontext.s.auto.html
+	 */
+	infoForControl(NULL, &info, OMRPORT_SIG_CONTROL_BP, &name, &reg);
+	unw_set_reg(cursor, UNW_X86_64_RBP, *(unw_word_t *)reg);
+	infoForControl(NULL, &info, OMRPORT_SIG_CONTROL_SP, &name, &reg);
+	/* Store call-site sp (skip return address) into RSP. */
+	unw_set_reg(cursor, UNW_X86_64_RSP, (*(unw_word_t *)reg) + 8);
+	infoForControl(NULL, &info, OMRPORT_SIG_CONTROL_PC, &name, &reg);
+	ip = *(unw_word_t *)reg;
+
+	if ((0 != size) && (nAddr < (size - 1))) {
```

Comment:
This condition is insufficient because at least two entries are added to the array. It should be
```
	if ((size >= 2) && (nAddr < (size - 2))) {
```

Reply:
Shouldn't the second condition stay the same since nAddr = 0 and size = 2 gives enough space for 2 entries?

Reply:
You're right. Perhaps what tripped me up was that the condition used `1`, yet was adding two (or more) entries. Changing `<` to `<=` would fix that mismatch.
```
	if ((size >= 2) && (nAddr <= (size - 2))) {
```